### PR TITLE
Add SearchAttributesManager

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
Add `SearchAttributesManager` with `GetSearchAttributes` and `SaveSearchAttributes` methods. Search attributes are cached and refreshed only once in specified interval (1 minute) if actively used.

**Why?**
These APIs will replace dynamic config search attributes in upcoming PRs.

**How did you test it?**
Existing and new unit tests.

**Potential risks**
No risks.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/clock/time_source.go [hunk 1]
```diff
@@ -45,7 +46,7 @@ type (
 
 	// EventTimeSource serves fake controlled time
 	EventTimeSource struct {
-		now time.Time
+		now atomic.Int64
```
Comment (alexshtin) [inline common/clock/time_source.go:49]: Had to switch this to atomic because of usage in concurrent tests.

File: common/cluster/metadata.go [hunk 2]
```diff
@@ -131,76 +163,155 @@ func NewMetadata(
...
+func (m *metadataImpl) GetSearchAttributes(indexName string, bypassCache bool) (map[string]enumspb.IndexedValueType, error) {
+	refreshCache := func(bypassCache bool, saCache searchAttributesCache, now time.Time) bool {
+		return bypassCache ||
+			saCache.lastRefresh.Add(searchAttributeCacheRefreshInterval).Before(now) ||
+			saCache.searchAttributes == nil
+	}
+
+	now := m.timeSource.Now()
+	saCache := m.searchAttributesCache.Load().(searchAttributesCache)
+
```
Comment (alexshtin) [inline common/cluster/metadata.go:253]: `GetSearchAttributes` and `SaveSearchAttributes` APIs are the core of this PR. They are not used yet but will replace dynamic config in upcoming PR.

It uses read optimized cache: `atomic` for read and `Mutex` for update.

File: common/cluster/metadata_test.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+					}}},
+		},
+		Version: 2,
+	}, nil)
+
+	wg := sync.WaitGroup{}
+	wg.Add(10)
+	for goroutine := 0; goroutine < 10; goroutine++ {
+		go func(goroutine int) {
+			for i := 1; i < 1500; i++ {
```
Comment (alexshtin) [inline common/cluster/metadata_test.go:123]: This tests concurrent read/writes and makes sure that DB read is called only expected number of times.

File: common/cluster/metadata.go [hunk 4]
```diff
@@ -56,8 +71,15 @@ type (
 		ClusterNameForFailoverVersion(failoverVersion int64) string
 	}
 
+	searchAttributesCache struct {
+		searchAttributes map[string]*persistencespb.IndexSearchAttributes
+		dbVersion        int64
+		lastRefresh      time.Time
+	}
```
Comment (wxing1292) [inline common/cluster/metadata.go:78]: search attribute management should belong to its own struct, not within cluster metadata

File: common/persistence/search_attributes_manager.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		searchAttributes = indexSearchAttributes.GetSearchAttributes()
+	}
+	if searchAttributes == nil {
+		return map[string]enumspb.IndexedValueType{}, nil
+	}
+	return searchAttributes, nil
+}
+
+// SaveSearchAttributes saves search attributes to cluster metadata.
+func (m *SearchAttributesManager) SaveSearchAttributes(indexName string, newCustomSearchAttributes map[string]enumspb.IndexedValueType) error {
```
Comment (wxing1292) [inline common/persistence/search_attributes_manager.go:124]: format

File: common/clock/time_source.go [hunk 6]
```diff
@@ -45,7 +46,7 @@ type (
 
 	// EventTimeSource serves fake controlled time
 	EventTimeSource struct {
-		now time.Time
+		now int64
```
Comment (wxing1292) [inline common/clock/time_source.go:49]: is EventTimeSource used in multi thread env?

File: common/persistence/search_attributes_manager.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+				saCache = searchAttributesCache{
+					searchAttributes: indexSearchAttributes,
+					lastRefresh:      now,
+					dbVersion:        clusterMetadata.Version,
+				}
+			} else {
+				saCache.lastRefresh = now
+			}
+			m.cache.Store(saCache)
+		}
```
Comment (wxing1292) [inline common/persistence/search_attributes_manager.go:109]: can we break this if into smaller functions?

**Review by wxing1292 — APPROVED** [review]

File: common/clock/time_source.go (hunk 6 above)
Comment (alexshtin) [reply common/clock/time_source.go:49]: Yes, this is why I have to make this conversion. Race detector complained about it.

File: common/persistence/search_attributes_manager.go (hunk 5 above)
Comment (alexshtin) [reply common/persistence/search_attributes_manager.go:124]: Do you want mulitline?

**Pull request merged**